### PR TITLE
DPE-2217 premptively switchover

### PR DESCRIPTION
**Pull request by paulomach**
## Issue

On scale down of primary, MySQL will reelect the primary after some delay.

## Solution

Minimize delay by preemptively setting primary to unit 0 (last one to be torn down) if primary is being removed.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -683,6 +683,16 @@ def _on_database_storage_detaching(self, _) -> None:
         if not self._mysql.is_instance_in_cluster(self.unit_label):
             return
 
+        # Preemptively switch primary to unit 0
+        if (
+            self._mysql.get_primary_label() == self.unit_label
+            and self.unit.name.split("/")[1] != "0"
+        ):
```
[the remark this answers is not in the document]
Comment (paulomach) [reply src/charm.py]: df0cd64

File: src/charm.py [hunk 2]
```diff
@@ -683,6 +683,16 @@ def _on_database_storage_detaching(self, _) -> None:
...
+            logger.debug("Switching primary to unit 0")
```
[the remark this answers is not in the document]
Comment (paulomach) [reply src/charm.py]: df0cd64

**Comment by paulomach** [comment]
> What happens if unit 0 is offline?

it would go boom. fixed on df0cd64

**Comment by paulomach** [comment]
> will this change remain k8s-only? if not, is it safe on vm—where having a unit 0 is not guaranteed?

VM need's another rule - commonly one would choose the leader unit. Eventually we merge the implementations differing the chosen unit as only difference

**Review by shayancanonical — APPROVED** [review]

**Pull request merged**